Sources/AsyncHTTPClient/HTTPHandler.swift

```diff
@@ -481,18 +481,65 @@ extension HTTPClient {
             let channel: Channel? = self.lock.withLock {
                 if !cancelled {
                     cancelled = true
-                    return self.channel
+                    return self.connection?.channel
                 }
                 return nil
             }
             channel?.triggerUserOutboundEvent(TaskCancelEvent(), promise: nil)
         }
 
         @discardableResult
-        func setChannel(_ channel: Channel) -> Channel {
+        func setConnection(_ connection: ConnectionPool.Connection) -> ConnectionPool.Connection {
             return self.lock.withLock {
-                self.channel = channel
-                return channel
+                self.connection = connection
+                return connection
+            }
+        }
+
+        func succeed(promise: EventLoopPromise<Response>?, with value: Response) {
+            if let connection = self.connection {
+                connection.channel.pipeline.removeHandler(name: "decompressHandler").recover { _ in }.flatMap {
+                    connection.channel.pipeline.removeHandler(name: "timeoutHandler")
+                }.recover { _ in }.flatMap { _ in
+                    connection.channel.pipeline.removeHandler(name: "taskHandler")
+                }.whenComplete { result in
+                    switch result {
+                    case .success:
+                        connection.release()
+                        promise?.succeed(value)
+                    case .failure(let error):
+                        fatalError("Couldn't remove taskHandler: \(error)")
+                    }
+                }
+            }
+        }
+
+        func fail(_ error: Error) {
+            if let connection = self.connection {
+                connection.channel.close().whenComplete { _ in
+                    connection.channel.pipeline.removeHandler(name: "taskHandler").whenComplete { result in
```

Comment:
Need to see if this method does the proper actions, is it actually needed to remove the task handler since we are closing the channel anyway?

Reply:
(Adding comments on my own PR to be sure not to merge before everything is fixed)

Reply:
@adtrevor closing the channel will anyway remove all handlers, no need to try to remove it again. Also if possible, prefer to not remove them by name.

Reply:
@weissi Thanks! Is the only alternative to keep a reference to remove them?
Is this because it is less type safe or for performance reasons?

Reply:
@adtrevor you can remove by type or reference. The reason not to use it name is that someone else might add one with the same name. In async-http-client however you control the whole pipeline so not that likely :)

Reply:
Thanks, done!